### DISCUSSION TITLE
-e should not imply --gtk-single-instance=false and --quit-after-last-window-closed=true

### DISCUSSION THREAD

**Discussion by SuperDuperDeou** [ghostty-org/ghostty#3880] [Feature Requests, Ideas] [answered]
`ghostty --gtk-single-instance=true -e foo` should make the existing instance spawn a new window and run `foo` in it. This would make behavior consistent with other single instance terminals in GNU/Linux (namely, Kitty and Wezterm). (Check #2978 )

A meta-flag (ex. `--force-single-instance`) that sets both `--gtk-single-instance=false` and `--quit-after-last-window-close=true` should be added, to be consistent with the other single instance terminals on GNU?Linux (ex. Wezterm's `--always-single-instance` and Kitty's `--wait-for-single-instance-window-close`, which isn't exactly the same thing, but has the same use case and does it better, IMHO).

**Comment by ruihe774** [comment]
> `ghostty --gtk-single-instance=true -e foo` should make the existing instance spawn a new window and run `foo` in it. This would make behavior consistent with other single instance terminals in GNU/Linux (namely, Kitty and Wezterm). (Check #2978 )

In single instance mode, a new ghostty process uses `g_application_activate` to communicate with the existing instance and let it to create a new window. The `activate` signal does not support passing arguments so currently it's not possible to tell the existing instance to spawn a command. A solution is to use `GActionEntry` to specify gtk actions e.g. "new-window", "new-tab", etc. and accept parameters in gtk actions. (I mean `GActionEntry`, not `cli.Action` nor `apprt.Action`)

Another problem is that configures like `command` and `initial-command` are stored globally. It is not possible to spawn a specific command without messing up the global config store, until these configures are made per-surface.

I don't think it is possible to support changing config switches like `ghostty --gtk-single-instance=true -e foo`. Config loading and parsing is the part of app initialization code. If the goal is to be quick, we should do it in `cli.Action.run` and do not initialize the app instance, just like other cli actions like "+list-fonts", "+show-config", etc. So I think it can be implemented in ways like `ghostty +new-tab --working-directory=bar -e foo` supporting a limited subset of configures that are made per-surface.

**Reply by SuperDuperDeou** [reply]
If think it this is done right, it's also a huge milestone towards supporting session management ( #3358 ), since a common use case is to have session scripts that set up a working environment (see Kitty's session files).

I am not opposed to it being implemented as an action. The problem with exec in single instance terminals is that sometimes you really want the process to not return until until the window is closed, and sometimes you just don't care. So, some toggle would be handy. `ghostty +window -e foo` should run foo in a new window and the process returns immediately, and `ghostty +window --wait-for-window-to-close -e foo` should run foo in a new window and the process returns only after the window is closed. The only problem is that that CLI would be janky with the upcoming `xdg-terminal-exec`, which assumes the exec arg to be one arg, not two.

Well, config being global rather than per surface impairs single instance mode support with `xdg-terminal-exec` in other ways, such as being unable to set a window class per window.

Kitty and Wezterm do not have that problem, it might be a good idea to look into how they're doing it.

**Reply by mitchellh** [reply]
> Kitty and Wezterm do not have that problem, it might be a good idea to look into how they're doing it.

They use their own daemon mode protocols, which makes this problem a whole lot easier.

**Comment by mitchellh** [comment, marked as answer]
This isn't currently planned. #2353 will likely help with this though.
